beacon_node/network/src/sync/block_sidecar_coupling.rs

```diff
@@ -265,6 +406,70 @@ impl<E: EthSpec> RangeBlockComponentsRequest<E> {
                     }
                 }
 
+                Some(resp)
+            }
+            // Reuse same logic that we use for coupling data columns for now.
+            // todo(pawan): we should never get a coupling error here, so simplify this
+            // variant's handling.
+            RangeBlockDataRequest::DataColumnsFromRoot {
+                all_requests_made,
+                attempt,
+                column_peers,
+                expected_custody_columns,
+                requests,
+            } => {
+                // Do not couple until requests covering all required columns
+                // have been made
+                if !*all_requests_made {
+                    return None;
+                }
+
+                let mut data_columns = vec![];
+                let mut column_to_peer_id: HashMap<u64, PeerId> = HashMap::new();
+                for req in requests.values() {
+                    let Some(data) = req.to_finished() else {
+                        return None;
+                    };
+                    data_columns.extend(data.clone())
+                }
+
+                // An "attempt" is complete here after we have received a response for all the
+                // requests we made. i.e. `req.to_finished()` returns Some for all requests.
+                *attempt += 1;
+
+                // Note: this assumes that only 1 peer is responsible for a column
+                // with a batch.
+                for (id, columns) in column_peers {
+                    for column in columns {
+                        column_to_peer_id.insert(*column, id.peer);
+                    }
+                }
+
+                let expected_custody_columns: Vec<_> =
+                    expected_custody_columns.iter().copied().collect();
+                let resp = Self::responses_with_custody_columns(
+                    blocks.to_vec(),
+                    data_columns,
+                    column_to_peer_id,
+                    &expected_custody_columns,
+                    *attempt,
+                );
+
+                if let Err(CouplingError::DataColumnPeerFailure {
+                    error: _,
+                    faulty_peers,
+                    action: _,
+                    exceeded_retries: _,
+                }) = &resp
+                {
+                    for (_, peer) in faulty_peers.iter() {
+                        // find the req id associated with the peer and
+                        // delete it from the entries as we are going to make
+                        // a separate attempt for those components.
+                        requests.retain(|&k, _| k.peer != *peer);
```

Comment:
should we also remove the same request from `column_peers` for consistency?
same here https://github.com/sigp/lighthouse/blob/e0d8f047ec13ca55c81be2b48d6b5f14bb53d6d1/beacon_node/network/src/sync/block_sidecar_coupling.rs#L405

Reply:
Good call.